odoo/addons/base/models/qweb.py

```diff
@@ -397,397 +244,311 @@ def get_template(self, template, options):
+                        if tokens[i]['type'] == token.NAME:
+                            allowed_keys.append(tokens[i]['string'])
+                        i += 1
+                elif string == 'for': # list comprehensions => allowed values for the current bracket depth
+                    i = index + 1
+                    while tokens[i] and (tokens[i]['type'] == token.OP or tokens[i]['type'] == token.NAME) and tokens[i]['string'] != 'in':
+                        if tokens[i]['type'] == token.NAME:
+                            allowed_keys.append(tokens[i]['string'])
+                        i += 1
+
+            index += 1
+
+
+        # nest bracket
+        index = 0
+        open_bracket_index = -1
+        bracket_depth = 0
+        while index < len(tokens):
+            t = tokens[index]
+            string = t['string']
+
+            if t['type'] == token.OP:
+                if string in ['(', '[', '{']:
+                    if bracket_depth == 0:
+                        open_bracket_index = index
+                    bracket_depth += 1
+                elif string in [')', ']', '}']:
+                    bracket_depth -= 1
+                    if bracket_depth == 0:
+                        code = self.__compile_expr_tokens(tokens[open_bracket_index + 1:index], list(allowed_keys), raise_on_missing)
+                        code = tokens[open_bracket_index]['string'] + code + tokens[index]['string']
+                        tokens = tokens[:open_bracket_index] + [{
+                            'string': code,
+                            'type': 'compiled',
+                            'start': tokens[open_bracket_index]['start'],
+                            'end': tokens[index]['end'],
+                        }] + tokens[index + 1:]
+                        index = open_bracket_index
+
+            index += 1
+
+        code = []
+        index = 0
+        pos = tokens and tokens[0]['start'] # to keep indent when use expr on multi line
+        while index < len(tokens):
+            t = tokens[index]
+            string = t['string']
+
+            if t['start'][0] != pos[0]:
+                pos = (t['start'][0], 0)
+            space = t['start'][1] - pos[1]
+            if space:
+                code.append(' ' * space)
+            pos = t['start']
+
+            if t['type'] == token.NAME:
+                if string == 'lambda': # lambda => allowed values
+                    code.append('lambda ')
+                    index += 1
+                    keys = []
+                    while tokens[index] and (tokens[index]['type'] == token.OP or tokens[index]['type'] == token.NAME) and tokens[index]['string'] != ':':
+                        if tokens[index]['type'] == token.NAME:
+                            key = tokens[index]['string']
+                            code.append(key)
+                            keys.append(key)
```

Comment:
Ah this is related to the one below, it doesn't distinguish between `NAME` which is a formal parameter name and `NAME` which is a value, so there really are two different holes:

* any `NAME` which is in the parameters list (including default values) is considered allowed_keys for the scope, meaning it won't be protected in the body e.g. `lambda a=globals: globals`
* independently, any `NAME` which is in the parameters list (including default values) is output as-is, meaning it will be bound to the parameter e.g. `lambda a=globals: a`

They're two independent issues.